tools/satsolver-main.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolutil/CDCL.h>
+#include <vector>
+#include <string>
+#include <iostream>
+#include <iomanip>
+#include <fstream>
+#include <sstream>
+
+using namespace solidity::util;
+using std::vector;
+using std::string;
+using std::cout;
+using std::endl;
+using std::optional;
+using std::nullopt;
+using std::make_pair;
+
+const int verbose = 0;
+vector<string> cutStringBySpace(string& input);
+optional<vector<Literal>> parseLine(std::string& line);
+std::pair<vector<vector<Literal>>, size_t> readCNFFile(const string& fname);
+size_t getNumVars(const vector<vector<Literal>>& cls);
+
+
+vector<string> cutStringBySpace(string& input)
+{
+    vector<string> parts;
+
+    size_t pos = 0;
+    while ((pos = input.find(" ")) != string::npos)
+	{
+        parts.push_back(input.substr(0, pos));
+        input.erase(0, pos + 1);
+    }
+    parts.push_back(input);
+    return parts;
+}
+
+optional<vector<Literal>> parseLine(std::string& line)
+{
+	vector<Literal> cl;
+	bool end_of_clause = false;
+	auto parts = cutStringBySpace(line);
+	for (const auto& part: parts)
```

Comment:
```suggestion
	for (const auto& part: line | ranges::views::split(' '))
```

Reply:
This is cool! Note, this requires `set_property(TARGET satsolver PROPERTY CXX_STANDARD 20)` in the `CMakeLIsts.txt` which may be a bit weird (to compile with a higher target than other binaries).

More importantly, it fails compilation at 

```
/home/matesoos/development/solidity/tools/satsolver-main.cpp:65:55: error:
no matching function for call to ‘std::__cxx11::basic_string<char>::basic_string(const std::ranges::split_view<std::ranges::ref_view<std::__cxx11::basic_string<char> >, std::ranges::single_view<char> >::_OuterIter<true>::value_type&)’
```

i.e.

```
const long lit = std::stol(part);
```

doesn't work. But I can't seem to fix this in a simple way :( Any pro tips to fix? I do like this `ranges` stuff, looks super-powerful, but somehow it doesn't spit out things that look like strings or chars. I tried `const long lit = std::stol(string(part));` but that doesn't work either. Do you have a pro tip to fix this?

Reply:
We use [range-v3](https://ericniebler.github.io/range-v3/) as polyfill until we can transition to C++20 - so you should just be able to link ``satsolver`` against the ``range-v3`` cmake target and include ``<range/v3/view/split.hpp>``. That'll also give you ``ranges::to`` in ``<range/v3/range/conversion.hpp>`` below (that's only staged for C++23, if at all).

Reply:
But unfortunately, ``part`` will be a range of characters, not a string, so indeed ``std::stol`` won't work...
```
std::from_chars(&*begin(part), &*begin(part) + ranges::distance(part), lit);
```
would work, but I admit that it's not really nice...

Reply:
Oh wow, that's great, fixes it perfectly, thanks! :bowing_man: Fixed in: 3822881f7bcbe38d6178216c30f848b7ecd8b25a

Reply:
OK, simplified in: cbbf3d572c4d4942381524810c706a664556cf1e Thanks! :)